Objects/genobject.c

```diff
@@ -642,6 +642,22 @@ _PyGen_SetStopIterationValue(PyObject *value)
         PyErr_SetObject(PyExc_StopIteration, value);
         return 0;
     }
+
+    // Since _PyGen_SetStopIterationValue() is meant to create
+    // a StopItertation or substitute one for a StopAsyncItertation
+    // an exception of another type should not already be set.
+    PyObject *run_ex = PyErr_Occurred();
+    if (run_ex) {
+        if (!PyErr_GivenExceptionMatches(run_ex, PyExc_StopAsyncIteration)) {
+            // Replace existing bad exception with a SystemError instead.
+            PyErr_Format(PyExc_SystemError,
+                         "%s:%d: unexpected caller exception: %R",
+                         __FILE__, __LINE__, run_ex);
```

Comment:
Do we have any precedent for passing `__FILE__` and `__LINE__` to errors from C? But anyways, I don't think displaying the exception like this is a good way to do it; we should just set it as a cause. (Also, this should probably be a `RuntimeError` instead of a `SystemError`--those should generally not be raisable from Python.)

You could do it with something like this:

```c
PyObject *cause = PyErr_GetRaisedException();
PyErr_SetString(PyExc_RuntimeError, "unexpected exception");
PyObject *err = PyErr_GetRaisedException();
PyException_SetCause(err, cause);
PyErr_SetRaisedException(err);
```

Reply:
We pass it when using BadInternalCall exception.

Reply:
> In addition I do want to crash the interpreter. if we are here, then the caller is wrong.

Oh! Maybe I'm misunderstanding the issue. The reproducer seems fairly simple--why should it raise a `SystemError`?

Reply:
Because I prefer that to having an `assert()`. I want release builds to report possible issues with this new construction (otherwise a plain `assert()` would be suppressed). WDYT?

Reply:
A `SystemError` is better than a crash, yeah, but in general we should avoid either of those if possible.  I see two other designs here: 

- Suppress all uncaught exceptions (either via `PyErr_Clear` or `PyErr_WriteUnraisable`)
- Raise them and don't try and set a `StopAsyncIteration`--if they weren't raised by a caller before, why not now?

Though, I'm not sure I fully understand the issue. Is the problem that an existing `StopAsyncIteration` can be set, and we're just speculating about other exceptions? Or can `_PyGen_SetStopIterationValue` be called with any arbitrary exception set?

Reply:
Ok, I'll try to explain here:

- We sometimes transform StopAsyncIteration into StopIteration exceptions. This is what happens in `anext`, e.g., `anext(it, default)`. If default is not a tuple, the fast path uses a plain `PyExc_SetObject()` and we're done. Otherwise, we fall back to a slow path but because we call `PyObject_CallOneArg()`, we cannot have an exception set. So we should clear it.
- Now, this is not always the case that `_PyGen_SetStopIterationValue` is called when we're already handling a StopAsyncIteration. I however couldn't find other call sites where we have another kind of exception set but still call `_PyGen_SetStopIterationValue`, which is why I'm raising a SystemError. Ideally, it would have been a `PyErr_BadInternalCall` but it wouldn't tell us the exception that was set and we would lose context for debugging.
- My check could actually be put *before* the fast path as well, but I wanted some confirmation beforehand. 

> Suppress all uncaught exceptions (either via PyErr_Clear or PyErr_WriteUnraisable)

I'm not sure it's not possible because we only want to suppress *some* of them (only StopAsyncIteration). On the other hand, we could clear the exception before calling `_PyGen_SetStopIterationValue` and assert that we don't have an exception set.

> Raise them and don't try and set a StopAsyncIteration--if they weren't raised by a caller before, why not now?

We must raise a StopIteration if a StopAsyncIteration was there (by the way, they are orthogonal classes). If we raise another exception that is not StopIteration, `anext` would raise as well.

Reply:
> PyErr_Occurred returns the exception type rather than the actual exception

Ah right I forgot. Well.. the rendering is still fine though but I think I'll go with `PyErr_BadInternalCall()` + setting a cause. That would be for the best.

Reply:
> I think we're doing too much with the hypothetical case where an exception that isn't a StopAsyncIteration is set

The problem is that it's an internal call and since we're still in an alpha release, I'd prefer that users give us as much logs as possible (which is only possible if we were to show sufficient information)l.